src/Configuration/Config.php

```diff
@@ -18,40 +21,109 @@ class Config
     /** @var PathResolver */
     private $pathResolver;
 
-    public function __construct()
+    /** @var Stopwatch */
+    private $stopwatch;
+
+    /** @var FilesystemCache */
+    private $cache;
+
+    public function __construct(Stopwatch $stopwatch)
     {
+        $this->stopwatch = $stopwatch;
+        $this->cache = new FilesystemCache();
         $this->pathResolver = new PathResolver(dirname(dirname(__DIR__)), []);
+        $this->data = $this->getConfig();
+    }
+
+    private function getConfig(): Collection
+    {
+        $this->stopwatch->start('bolt.parseconfig');
+
+        if ($this->validCache()) {
+            $data = $this->getCache();
+        } else {
+            $data = $this->parseConfig();
+            $this->setCache($data);
+        }
+
+        $this->stopwatch->stop('bolt.parseconfig');
 
-        $this->data = $this->parseConfig();
+        return $data;
+    }
+
+    private function validCache(): bool
+    {
+        if (! $this->cache->has('config_cache') || ! $this->cache->has('config_timestamps')) {
+            return false;
+        }
+
+        $timestamps = $this->cache->get('config_timestamps');
+
+        foreach ($timestamps as $filename => $timestamp) {
+            if (filemtime($filename) > $timestamp) {
+                return false;
```

Comment:
IMO cache should store each config file separately and the file has been modified and then just set cache->expiresAt() as PSR-6 says

Reply:
@JarJak That won't work in this case, if only for the example pointed out by Marcin yesterday: Let's say the `.env` file changes. In that case not only _that_ cached part should expire, but the entire cached config. 

In this case, i think it'll also make it quite. bit more complex. Let's do that only if we're a bit further along, and we actually measure there's performance to be gained here. :-)